crates/exex/exex/src/wal/storage.rs

```diff
@@ -0,0 +1,200 @@
+    /// it doesn't exist.
+    pub(super) fn new(path: impl AsRef<Path>) -> eyre::Result<Self> {
+        reth_fs_util::create_dir_all(&path)?;
+
+        let (mut min_id, mut max_id) = (None, None);
+
+        for entry in reth_fs_util::read_dir(&path)? {
+            let entry = entry?;
+            let file_name = entry.file_name();
+            let file_id = Self::parse_filename(&file_name.to_string_lossy())?;
+
+            min_id = min_id.map_or(Some(file_id), |min_id: u64| Some(min_id.min(file_id)));
+            max_id = max_id.map_or(Some(file_id), |max_id: u64| Some(max_id.max(file_id)));
+        }
+
+        debug!(?min_id, ?max_id, "Initialized WAL storage");
+
+        Ok(Self { path: path.as_ref().to_path_buf(), min_id, max_id })
+    }
+
+    fn file_path(&self, id: u64) -> PathBuf {
+        self.path.join(format!("{id}.wal"))
+    }
+
+    fn parse_filename(filename: &str) -> eyre::Result<u64> {
+        filename
+            .strip_suffix(".wal")
+            .and_then(|s| s.parse().ok())
+            .ok_or_eyre(format!("failed to parse file name: {filename}"))
+    }
+
+    fn adjust_file_range(&self, range: impl RangeBounds<u64>) -> Option<RangeInclusive<u64>> {
+        let (min_id, max_id) = self.min_id.zip(self.max_id)?;
+
+        let start = match range.start_bound() {
+            Bound::Included(start) => *start,
+            Bound::Excluded(start) => *start + 1,
+            Bound::Unbounded => min_id,
+        };
+        let end = match range.end_bound() {
+            Bound::Included(end) => *end,
+            Bound::Excluded(end) => *end - 1,
+            Bound::Unbounded => max_id,
+        };
+
+        Some(start..=end)
+    }
+
+    /// Removes notifications from the storage according to the given range.
+    pub(super) fn remove_notifications(
+        &mut self,
+        range: RemoveNotificationsRange,
+    ) -> eyre::Result<Vec<ExExNotification>> {
+        let adjusted_range = match range {
+            RemoveNotificationsRange::FromFileId(from_file_id) => {
+                self.adjust_file_range(from_file_id..)
+            }
+            RemoveNotificationsRange::ToFileId(to_file_id) => self.adjust_file_range(..to_file_id),
+        };
+        let Some(adjusted_range) = adjusted_range else { return Ok(Vec::new()) };
+
+        let removed_notifications =
+            self.iter_notifications(adjusted_range).collect::<eyre::Result<Vec<_>>>()?;
+
+        for (id, _) in &removed_notifications {
+            debug!(?id, "Removing notification from the storage");
+            reth_fs_util::remove_file(self.file_path(*id))?;
+        }
+
+        match range {
+            RemoveNotificationsRange::FromFileId(from_file_id) => {
+                self.max_id = from_file_id.checked_sub(1)
+            }
+            RemoveNotificationsRange::ToFileId(to_file_id) => self.min_id = Some(to_file_id),
+        };
+
+        Ok(removed_notifications.into_iter().map(|(_, notification)| notification).collect())
```

Comment:
why do we need this after we deleted the file?

Reply:
for the rollback: https://github.com/paradigmxyz/reth/blob/67f8c1d1a40d28b9dd2f127f4d366c55a60967de/crates/exex/exex/src/wal/mod.rs#L22-L25

for the finalization, there's no reason and I will remove it, so that we decode and collect notifications only on rollback

Reply:
i'm open to unifying these methods too, but note that we can't just have a lazy iterator because we need to delete the files immediately